consistent/consistent.go

```diff
@@ -0,0 +1,271 @@
+// get remapped.
+//
+// Read more about consistent hashing on wikipedia:  http://en.wikipedia.org/wiki/Consistent_hashing
+//
+package consistent
+
+import (
+	"errors"
+	"hash/fnv"
+	"sort"
+	"strconv"
+	"sync"
+
+	. "github.com/thunderdb/ThunderDB/proto"
+)
+
+type NodeKeys []NodeKey
+
+// Len returns the length of the uints array.
+func (x NodeKeys) Len() int { return len(x) }
+
+// Less returns true if node i is less than node j.
+func (x NodeKeys) Less(i, j int) bool { return x[i] < x[j] }
+
+// Swap exchanges nodes i and j.
+func (x NodeKeys) Swap(i, j int) { x[i], x[j] = x[j], x[i] }
+
+// ErrEmptyCircle is the error returned when trying to get an node when nothing has been added to hash.
+var ErrEmptyCircle = errors.New("empty circle")
+
+// Consistent holds the information about the members of the consistent hash circle.
+type Consistent struct {
+	circle           map[NodeKey]Node
+	members          map[NodeId]Node
+	sortedHashes     NodeKeys
+	NumberOfReplicas int
+	count            int64
+	scratch          [64]byte
+	sync.RWMutex
+}
+
+// New creates a new Consistent object with a default setting of 20 replicas for each entry.
+//
+// To change the number of replicas, set NumberOfReplicas before adding entries.
+func New() *Consistent {
+	c := new(Consistent)
+	c.NumberOfReplicas = 20
+	c.circle = make(map[NodeKey]Node)
+	c.members = make(map[NodeId]Node)
+	return c
+}
+
+// nodeKey generates a string key for an node with an index.
+func (c *Consistent) nodeKey(node_id NodeId, idx int) string {
+	// return node + "|" + strconv.Itoa(idx)
+	return strconv.Itoa(idx) + string(node_id)
+}
+
+// Add inserts a string node in the consistent hash.
+func (c *Consistent) Add(node Node) {
+	c.Lock()
+	defer c.Unlock()
+	c.add(node)
+}
+
+// need c.Lock() before calling
+func (c *Consistent) add(node Node) {
+	for i := 0; i < c.NumberOfReplicas; i++ {
+		c.circle[c.hashKey(c.nodeKey(node.Id, i))] = node
+	}
+	c.members[node.Id] = node
+	c.updateSortedHashes()
+	c.count++
+}
+
+// Remove removes an node from the hash.
+func (c *Consistent) Remove(node Node) {
+	c.Lock()
+	defer c.Unlock()
+	c.remove(node.Id)
+}
+
+// need c.Lock() before calling
+func (c *Consistent) remove(node_id NodeId) {
+	for i := 0; i < c.NumberOfReplicas; i++ {
+		delete(c.circle, c.hashKey(c.nodeKey(node_id, i)))
+	}
+	delete(c.members, node_id)
+	c.updateSortedHashes()
+	c.count--
+}
+
+// Set sets all the nodes in the hash.  If there are existing nodes not
+// present in nodes, they will be removed.
+func (c *Consistent) Set(nodes []Node) {
+	c.Lock()
+	defer c.Unlock()
+	for k := range c.members {
+		found := false
+		for _, v := range nodes {
+			if k == v.Id {
+				found = true
+				break
+			}
+		}
+		if !found {
+			c.remove(k)
+		}
+	}
+	for _, v := range nodes {
+		_, exists := c.members[v.Id]
+		if exists {
+			continue
+		}
+		c.add(v)
+	}
+}
+
+func (c *Consistent) Members() []Node {
+	c.RLock()
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported method Consistent.Members should be of the form "Members ..."